jest/react-native-env.js

```diff
@@ -0,0 +1,16 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @format
+ */
+
+'use strict';
+
+const NodeEnv = require('jest-environment-node').TestEnvironment;
```

Comment:
Should add a dependency on `jest-environment-node` - otherwise you rely on hoisting.

When you stop extending the existing env, that dep can of course be removed.

Reply:
sounds good - originally I added it to the repo-config package.json, but since it didn't change the `yarn.lock` I assumed it was part of the dependencies that jest itself was bringing along. Will modify PR accordingly 👍

Reply:
> since it didn't change the `yarn.lock` I assumed it was part of the dependencies that jest itself was bringing along

that's correct, but it would then rely on hoisting 🙂 

It needs to be part of the root package.json btw - same level as the preset lives on

Reply:
> It needs to be part of the root package.json btw - same level as the preset lives on. Which might introduce issues since there's no jest there... Possibly an optional peer dep makes more sense than a direct dependency? Again, will be removed when you stop extending.

yeah the thing is that the repo-config/package.json is this really weird thing (the plan is to remove it)... goes back to the monorepo conversation here https://github.com/react-native-community/discussions-and-proposals/pull/480

I'm going to guess this dep should live alongside Jest, and that means that I'll have to add it to the repo-config one and then modify the magic script that does some stuff when branch cutting to ensure that on released versions it gets moved along (you can see what I mean here https://github.com/facebook/react-native/commit/c0127265a15f6d01217d33c9fab4750a27e40f3c#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519)

Reply:
That won't hit same issue as https://github.com/facebook/react-native/pull/31668?

Reply:
to be on the side of caution I ended up adding it as `dependency` to root - basically the repo-config dependencies get copied over here https://github.com/facebook/react-native/blob/main/scripts/set-rn-version.js#L114 but they end up in the  `devDependency` group so that might be why the issue you linked was happening.

Reply:
If/when you get a monorepo with multiple packages going, separating the preset into its own package with its own deps probably makes sense 🙂

Reply:
> to be on the side of caution I ended up adding it as `dependency` to root

Yeah, I think this is right (/our best option) for the `react-native` monolith - `jest-environment-node` should be a `dependency` of `react-native`, because `react-native` provides the preset (and environment).

> (it might make sense to just copy the entire content of jest-environment-node now instead of empty extend? That way, no dep)

I'm in favour of this as a next step, but `jest-environment-node` has other dependencies on Jest so it'd only move this particular problem.

Reply:
Very true 😅 separate package for the preset entirely is probably the least wrong approach (longer term - short term this should work a treat of course)

Reply:
right, almost forgot - added a "next steps" section to the PR body with all the things that should be done as follow ups :)

Reply:
I feel like I mentioned it but I can't find it so might've been a fever dream.

An option to a dep is optional peer dep. Sorta depends on timeline for 0.71 - seems suboptimal for `yarn add react-native` to bundle a jest env.

Reply:
> An option to a dep is optional peer dep. Sorta depends on timeline for 0.71 - seems suboptimal for `yarn add react-native` to bundle a jest env.

Fair point. Correct me if I'm wrong - that would also mean adding `jest-node-environment` to the template's `devDependencies` explicitly (alongside `jest`)? I don't *think* peer dependencies are guaranteed to be hoisted if they're only provided transitively (via `jest` in this case), I think they're expected to be specified by the root project. 

We'd have to introduce `peerDependenciesMeta` also, since `react-native` doesn't currently have any optional peer deps.

I'm ok with the current `dependencies` approach for a few reasons;
 - simplicity, especially re upgrade paths for userland `package.json`,
 - `jest-environment-node` is small,
 - we're moving towards a monorepo,
 - we plan to drop the dependency on `jest-environment-node`
 
 @cortinico, @hramos , @motiz88 - want to weigh in here?

Reply:
(for context: we are aiming to cut 0.71 over the next couple weeks)